src/vs/editor/contrib/referenceSearch/browser/referencesModel.ts

```diff
@@ -201,9 +208,20 @@ export class ReferencesModel implements IDisposable {
 	private _references: OneReference[] = [];
 	private _eventBus = new EventEmitter();
 
-	onDidChangeReferenceRange: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'ref/changed');
+	onDidChangeReferenceRange: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'range/changed');
+	onDidChangeReferences: Event<OneReference> = fromEventEmitter<OneReference>(this._eventBus, 'refs/changed');
 
 	constructor(references: Location[]) {
+		this._setReferences(references, false);
+	}
+
+	public setReferences(references: Location[]) {
+		this._setReferences(references, true);
```

Comment:
I understand the concern that you have, but I disagree on some points.

> It should not push down C because the user would then select B instead of C.

You are of course theoretically correct. My arguments here are practical:
1. It doesn't happen that often in practice
2. Clicking on the wrong thing doesn't cost the user anything (UI doesn't change in a dramatic way, can just click again on the intended result)
3. We could throttle updates to predictable intervals (e.g every 1s) so the user can intuitively know when an update might happen.
4. The user can just wait for all results to return (which is what they are forced to do now) before interacting. They are strictly better off with streaming.
5. As long as the user's current selection/view is maintained, I think it is ok for the results around it to be updating.

> It will push down the C-subtree, making the user click the wrong item.

Only if subtree A is expanded. I forget what the behavior is in the PR as it stands now, but we wouldn't need to auto expand any subtree automatically until all results are in. I think the auto expand behavior depends on # results in a subtree so it wouldn't make sense to auto expand anyway until all results are in.

> instead of resetting and resorting all locations we should always have append-insert semantics.

I got the opposite feedback on https://github.com/Microsoft/vscode/pull/28751, because the sorting was relevant there. As of right now, I would agree that the sorting isn't as relevant for references here (still a non-zero amount of relevant), but we would be precluding it from being more relevant in the future.

It also seems a little bit inconsistent if we have different behaviors for streaming in different contexts.

---

I am not sure if I will have time to make progress on this in the next two weeks.